### DISCUSSION TITLE
NODE_OPTIONS compilation warning after upgrading to 7.x

### DISCUSSION THREAD

**Discussion by robotfelix** [shakacode/shakapacker#488] [Q&A] [answered]
After upgrading shakapacker from `6.6.0` to `7.2.3`, I'm now getting warnings about conflicting values of `NODE_OPTIONS`
```
WARNING in DefinePlugin
Conflicting values for 'process.env.NODE_OPTIONS'
```

I've tested and found:
* `NODE_OPTIONS="--trace-warnings"` in `.env` and calling `./bin/shakapacker-dev-server` **=> warning**
* No `NODE_OPTIONS` in `.env` and calling `./bin/shakapacker-dev-server` **=> no warning**
* `NODE_OPTIONS="--trace-warnings"` in `.env` and calling `NODE_OPTIONS="--trace-warnings" ./bin/shakapacker-dev-server` **=> no warning**

So I think the changes made in #350 are conflicting with the `NODE_OPTIONS` configuration I have in a dotenv / `.env` file and setting the value of `NODE_OPTIONS` before the .env file is be read.

Ideally I wouldn't have to stop using the .env for configuring this value as it varies between projects.

Has anyone else encountered a similar problem and has a suggestion of how to fix it?

**Comment by tomdracz** [comment]
How are you passing the values from .env file to your app?

Are you able to provide your webpack config? Can investigate further if we have a reproduction case

**Reply by robotfelix** [reply]
We have the `dotenv-webpack` as a dependency in our `package.json`, as well as `dotenv-rails` in our `Gemfile`.

I think the relevant bits of our webpack config are these:
```js
// my-config.js
const webpack = require('webpack');
const Dotenv = require('dotenv-webpack');
// ...
const {generateWebpackConfig, inliningCss} = require('shakapacker');
const ReactRefreshWebpackPlugin = require('@pmmmwh/react-refresh-webpack-plugin');

module.exports = {
  // ...
  plugins: [
    // ...
    new Dotenv(),
    // ...
  ].filter(Boolean),
  // ...
};
```

```js
// webpack.config.js
const {generateWebpackConfig, merge} = require('shakapacker')
const myConfig = require('./my-config')
const webpackConfig = generateWebpackConfig()

module.exports = merge(webpackConfig, myConfig)
```

I think `Shakapacker::DevServerRunner` will initialise `NODE_OPTIONS` to an empty string before the JS code does anything with the ENV though.

**Reply by Judahmeek** [reply, marked as answer]
The cause of the error is likely that the `Dotenv` plugin calls `DefinePlugin` & so does [the EnvironmentPlugin in Shakapacker's default webpack configuration](https://github.com/shakacode/shakapacker/blob/main/package/plugins/webpack.ts#L18).

The easiest solution is to overwrite the `plugins` property, instead of merging.
